### PR TITLE
Add dynamic scale

### PR DESCRIPTION
**Pull request by yhcao6**
move https://github.com/open-mmlab/mmdetection/pull/3723 to mmcv

### CODE REVIEW

**Comment by hellock** [comment]
Task linked: [CU-4qrg3x [In Review] Dynamic Loss Scale](https://app.clickup.com/t/4qrg3x)

**Review by RyanXLi — APPROVED** [review]

**Review by RyanXLi — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

File: mmcv/runner/fp16_utils.py [hunk 1]
```diff
@@ -264,3 +264,84 @@ def new_forward(*args, **kwargs):
...
+    Dynamic loss scaling begins by attempting a very high loss
+    scale.  Ironically, this may result in OVERflowing gradients.
+    If overflowing gradients are encountered, :class:`FP16_Optimizer` then
+    skips the update step for this particular iteration/minibatch,
+    and :class:`LossScaler` adjusts the loss scale to a lower value.
+    If a certain number of iterations occur without overflowing gradients
+    detected,:class:`LossScaler` increases the loss scale once more.
+    In this way :class:`LossScaler` attempts to "ride the edge" of always
+    using the highest loss scale possible without incurring overflow.
+    Args:
```
Comment (ZwwWayne) [inline mmcv/runner/fp16_utils.py:288]: Need an empty line in docstring above Args.

File: mmcv/runner/fp16_utils.py [hunk 2]
```diff
@@ -264,3 +264,84 @@ def new_forward(*args, **kwargs):
...
+    """
+
+    def __init__(self,
+                 init_scale=2**32,
+                 mode='dynamic',
+                 scale_factor=2.,
+                 scale_window=1000):
+        self.cur_scale = init_scale
+        self.cur_iter = 0
+        self.mode = mode
```
Comment (hellock) [inline mmcv/runner/fp16_utils.py:304]: Validate `mode`.

File: mmcv/runner/fp16_utils.py [hunk 3]
```diff
@@ -264,3 +264,87 @@ def new_forward(*args, **kwargs):
...
+        scale_factor (float): Factor used when adjusting the loss scale.
+            Default: 2.
+        mode (str): Loss scaling mode. 'dynamic' or 'static'
+        scale_window (int): Number of consecutive iterations without an
+            overflow to wait before increasing the loss scale. Default: 1000.
+    """
+
+    def __init__(self,
+                 init_scale=2**32,
+                 mode='dynamic',
```
Comment (hellock) [inline mmcv/runner/fp16_utils.py:300]: This changes the default behavior.

**Pull request merged**